apps/condo/domains/organization/utils/accessSchema.js

```diff
@@ -1,239 +1,308 @@
+        id_in: userRoleIds,
+        // TODO(SavelevMatthew): INFRA-317 Add this check
+        // deletedAt: null,
+    })
 
-    for (const id of organizationIds) {
-        const hasAccess = await checkOrganizationPermission(userId, id, permission)
-        if (!hasAccess) return false
+    for (const role of userRoles) {
+        newCacheEntry.organizations[role.organization] = {
+            permissions: _extractRolePermissions(role),
+            childOrganizations: [],
+        }
     }
 
-    return true
+    const organizationLinks = await find('OrganizationLink', {
+        from: { id_in: Object.keys(newCacheEntry.organizations) },
+        deletedAt: null,
+        to: { deletedAt: null },
+    })
+
+    for (const link of organizationLinks) {
+        newCacheEntry.organizations[link.from].childOrganizations.push(link.to)
+    }
+
+    if (!DISABLE_USER_ORGANIZATION_CACHING) {
+        await _redisClient.set(cacheKey, JSON.stringify(newCacheEntry), 'PX', CACHE_TTL_IN_MS)
+    }
+
+    return newCacheEntry
 }
 
 /**
- * Check that the user has access in each related organization
- *
- * @param userId {string} User.id field
- * @param organizationIds {Array<string>} array of objects related organizations
- * @param permission {string} OrganizationEmployeeRole permission key to check for
- * @return {Promise<boolean>}
+ * Checks if user have all specified permissions in organizations
+ * @param {Record<string, boolean>} organizationPermissions
+ * @param {Array<string>} permissionsToCheck
+ * @private
  */
-async function checkRelatedOrganizationsPermission (userId, organizationIds, permission) {
-    if (!userId || !isArray(organizationIds) || isEmpty(organizationIds)) return false
+function _checkPermissionsInOrganization (organizationPermissions, permissionsToCheck) {
+    return permissionsToCheck.every(permission => organizationPermissions.hasOwnProperty(permission) && organizationPermissions[permission])
+}
 
-    const uniqOrganizationIds = uniq(organizationIds)
+/**
+ * Gets the IDs of organizations where user is employed and its employee has all permissions from the list
+ * @param {{ id: string }} user - user object
+ * @param {Array<string> | string} permissions - permissions to check,
+ * can be passed as array of strings for multiple permissions or a single string for a single permission
+ * @returns {Promise<Array<string>>}
+ */
+async function getEmployedOrganizationsByPermissions (user, permissions) {
+    const userOrganizationsInfo = await _getUserOrganizations(user)
 
-    if (!uniqOrganizationIds.every(isUUID)) return false
+    const permissionsToCheck = Array.isArray(permissions) ? permissions : [permissions]
+    const organizationIds = []
 
-    for (const id of organizationIds) {
-        const hasAccess = await checkRelatedOrganizationPermission(userId, id, permission)
-        if (!hasAccess) return false
+    for (const [id, info] of Object.entries(userOrganizationsInfo.organizations)) {
```

Comment:
Probably at beginning of function it is worth checking that `user `was transferred, and if they are empty, then return false.

This logic was in previous utilities, and we won’t have to request data in vain